src/backend/common/CommandBufferStateTracker.cpp

```diff
@@ -0,0 +1,455 @@
+        }
+        return ValidateCanDraw();
+    }
+
+    bool CommandBufferStateTracker::BeginSubpass() {
+        auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
+        for (auto location : IterateBitSet(subpassInfo.colorAttachmentsSet)) {
+            auto attachmentSlot = subpassInfo.colorAttachments[location];
+            auto* tv = currentFramebuffer->GetTextureView(attachmentSlot);
+            // TODO(kainino@chromium.org): the TextureView can only be null
+            // because of the null=backbuffer hack (null representing the
+            // backbuffer). Once that hack is removed (once we have WSI)
+            // this check isn't needed.
+            if (tv == nullptr) {
+                continue;
+            }
+
+            auto* texture = tv->GetTexture();
+            if (texture->HasFrozenUsage(nxt::TextureUsageBit::ColorAttachment)) {
+                continue;
+            }
+            if (!texture->IsTransitionPossible(nxt::TextureUsageBit::ColorAttachment)) {
+                builder->HandleError("Can't transition attachment to ColorAttachment usage");
+                return false;
+            }
+            mostRecentTextureUsages[texture] = nxt::TextureUsageBit::ColorAttachment;
+            texturesTransitioned.insert(texture);
+        }
+        return true;
+    };
+
+    void CommandBufferStateTracker::EndSubpass() {
+        auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
+        for (auto location : IterateBitSet(subpassInfo.colorAttachmentsSet)) {
+            auto attachmentSlot = subpassInfo.colorAttachments[location];
+            auto* tv = currentFramebuffer->GetTextureView(attachmentSlot);
+            // TODO(kainino@chromium.org): the TextureView can only be null
+            // because of the null=backbuffer hack (null representing the
+            // backbuffer). Once that hack is removed (once we have WSI)
+            // this check isn't needed.
+            if (tv == nullptr) {
+                continue;
+            }
+
+            auto* texture = tv->GetTexture();
+            if (texture->IsFrozen()) {
+                continue;
+            }
+            mostRecentTextureUsages[texture] = nxt::TextureUsageBit::None;
+        }
+    };
+
+    bool CommandBufferStateTracker::BeginRenderPass(RenderPassBase* renderPass, FramebufferBase* framebuffer) {
+        if (currentRenderPass != nullptr) {
+            builder->HandleError("A render pass is already active");
+            return false;
+        }
+        if (!framebuffer->GetRenderPass()->IsCompatibleWith(renderPass)) {
+            builder->HandleError("Framebuffer is incompatible with this render pass");
+            return false;
+        }
+
+        currentRenderPass = renderPass;
+        currentFramebuffer = framebuffer;
+        currentSubpass = 0;
+        if (!BeginSubpass()) {
+            return false;
+        }
+
+        aspects.reset(VALIDATION_ASPECT_COMPUTE_PIPELINE);
```

Comment:
note: I don't know how inheritance of bindings and vertices works at renderpass start in the backing APIs, but it seems likely that we will have to reset all aspects here. This way applications will be forced to set things again in the subpass, which will remove the need to do some state-tracking in the backends.